server/src/main/java/org/elasticsearch/action/admin/indices/template/post/TransportSimulateIndexTemplateAction.java

```diff
@@ -0,0 +1,176 @@
+                                   ActionListener<SimulateIndexTemplateResponse> listener) throws Exception {
+        ClusterState simulateOnClusterState = state;
+        if (request.getIndexTemplateRequest() != null) {
+            // we'll "locally" add the template defined by the user in the cluster state (as if it existed in the system)
+            String simulateTemplateToAdd = "simulate_new_template_" + UUIDs.randomBase64UUID();
+            simulateOnClusterState = indexTemplateService.addIndexTemplateV2(state, request.getIndexTemplateRequest().create(),
+                simulateTemplateToAdd, request.getIndexTemplateRequest().indexTemplate());
+        }
+
+        String matchingTemplate = findV2Template(simulateOnClusterState.metadata(), request.getIndexName(), false);
+        if (matchingTemplate == null) {
+            listener.onResponse(new SimulateIndexTemplateResponse(null, null));
+            return;
+        }
+        Settings settings = resolveSettings(simulateOnClusterState.metadata(), matchingTemplate);
+
+        // empty request mapping as the user can't specify any explicit mappings via the simulate api
+        Map<String, Object> mappings = resolveV2Mappings("{}", simulateOnClusterState, matchingTemplate, xContentRegistry);
+        String mappingsJson = Strings.toString(XContentFactory.jsonBuilder()
+            .startObject()
+            .field(MapperService.SINGLE_MAPPING_NAME, mappings)
+            .endObject());
+
+        List<Map<String, AliasMetadata>> resolvedAliases = MetadataIndexTemplateService.resolveAliases(simulateOnClusterState.metadata(),
+            matchingTemplate);
+
+        // create the index with dummy settings in the cluster state so we can parse and validate the aliases
+        Settings dummySettings = Settings.builder()
+            .put(IndexMetadata.SETTING_VERSION_CREATED, Version.CURRENT)
+            .put(settings)
+            .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
+            .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
+            .put(IndexMetadata.SETTING_INDEX_UUID, UUIDs.randomBase64UUID())
+            .build();
+        final IndexMetadata indexMetadata = IndexMetadata.builder(request.getIndexName()).settings(dummySettings).build();
+
+        simulateOnClusterState = ClusterState.builder(simulateOnClusterState)
+            .metadata(Metadata.builder(simulateOnClusterState.metadata())
+                .put(indexMetadata, true)
+                .build())
+            .build();
+
+        IndexService tempIndexService = indicesService.createIndex(indexMetadata, Collections.emptyList(), false);
+        final Index index = tempIndexService.index();
+        try (Closeable dummy = () -> tempIndexService.close("temp", false)) {
+            List<AliasMetadata> aliases = MetadataCreateIndexService.resolveAndValidateAliases(request.getIndexName(), Set.of(),
```

Comment:
In subsequent work (doesn't have to be this PR) maybe we can extract this into a helper method, creating an IndexService for the simulation here makes me a little uneasy for a long-term solution.

Reply:
completely agree, this is clunky and rather error-prone (easy to forget to close it etc)